content/rest-api/good-practices.md

```diff
@@ -0,0 +1,98 @@
+
+   Cache frequently accessed data on the client side to reduce repetitive API requests and improve response times.
+
+## Retry strategy
+
+When integrating with any external API, it's important to anticipate and gracefully handle potential issues. While we strive for high availability, there are rare occasions where you might encounter server-side errors or network instability.
+
+The following responses indicate transient problems that are useful to retry:
+
+- HTTP 408, 429, and 5xx response codes.
+- Socket timeouts and TCP disconnects.
+
+
+**Understanding 5xx Errors**
+
+   A 5xx status code indicates that the server encountered an unexpected condition that prevented it from fulfilling the request. These errors are usually transient, meaning they are often resolved quickly without any action required on your part beyond a retry. Building your system to anticipate and gracefully handle these temporary server-side issues is crucial for a resilient integration.
+
+**Good Practices for Retries**
+
+   When faced with a 5xx error, the recommended approach is to implement a retry mechanism. This involves making the same API request again after a short delay. However, simply retrying immediately can exacerbate issues if the server is truly overloaded. Here are some best practices for implementing intelligent retries:
+
+1. **Exponential Backoff with Jitter**
+
+      This is the most critical principle for retries. Instead of retrying after a fixed delay, you should increase the delay exponentially between successive retries. This prevents overwhelming our servers with repeated requests during a period of instability and gives the server time to recover. To avoid a "thundering herd" problem where many clients retry at the exact same moment, introduce a small, random amount of "jitter" to your backoff times.
+
+      How it works:
+
+      Initial Delay: Start with a small delay (e.g., 100ms, 500ms, or 1 second).
+
+      Subsequent Delays: Double or exponentially increase the delay for each subsequent retry. For example: 1s, 2s, 4s, 8s, 16s, etc.
+
+      Jitter: Add a small, random deviation to your calculated delay.
+
+      Example Exponential Backoff with Jitter (Conceptual):
+      ```
+      retry_delay = initial_delay
+      for attempt in range(max_retries):
+         try:
+            make_api_request()
+            break // Success, exit loop
+         except (5xx error):
+            sleep(retry_delay + random_jitter)
+            retry_delay *= backoff_factor // e.g., 2
+      ```
+
+2. **Maximum Number of Retries**
+
+   Define a reasonable maximum number of retry attempts. Continuously retrying indefinitely can lead to resource exhaustion on your end and unnecessary load on our servers. Typically, 3 to 5 retries with exponential backoff are sufficient for most transient issues. If the error persists after the maximum retries, it's likely a more significant problem that requires alerting and manual intervention.
```

Comment:
We talk about 3 to 5 retries, meaning the delay between the first and the last retry will be between 1+2+4=7s and 1+2+4+8+16=31s.
Is this enough to avoid definitive failures?

Recommending up to 8 retries leads to 255 seconds between first and last retry.

I'd recommend between 5 and 8 retries, rather than between 3 and 5.